### PR TITLE
Implement ungridded dim mirroring and complete ungridded dim parsing

### LINKED ISSUE #2900: Add support for "mirror" of ungridded dims in MAPL3
The default should be *no* ungridded dims, but we want an option to specify that the ungridded dims of a FieldSpec should come from whatever field to which it is ultimately connected.

In this case, only imports should be allowed to specify mirror.   But for now we will allow exports.    Other mirror things make sense for ExtData, but not this one.

@bena-nasa @darianboggs

### PR DESCRIPTION
**Pull request by bena-nasa**
Fixes #2900 
This PR implements mirroring of ungridded dims in MAPL3. In addition it also completes the parsing of the ungridded dim from the yaml file since that had not been completed; only the extent had actually been read from the file,  the name, the units, if the user specified coordinates rather than the extent, none of that was being parsed. 

I did change the constructor for the ungriddedDim class so it would be easier to use when parsing the yaml file.

## Types of change(s)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist
- [ ] Tested this change with a run of GEOSgcm
- [ ] Ran the Unit Tests (`make tests`)

### CODE REVIEW

File: generic3g/ComponentSpecParser/parse_var_specs.F90 [hunk 1]
```diff
@@ -223,15 +224,31 @@ function to_UngriddedDims(attributes,rc) result(ungridded_dims)
...
-            dim_size = ESMF_HConfigAsI4(dim_spec, keyString=KEY_UNGRIDDED_DIM_EXTENT, _RC)
-            temp_dim = UngriddedDim(dim_size)
+            has_name = ESMF_HConfigIsDefined(dim_spec,keyString=KEY_UNGRIDDED_DIM_NAME)
+            has_units = ESMF_HConfigIsDefined(dim_spec,keyString=KEY_UNGRIDDED_DIM_UNITS)
+            has_extent = ESMF_HConfigIsDefined(dim_spec,keyString=KEY_UNGRIDDED_DIM_EXTENT)
+            has_coordinates = ESMF_HConfigIsDefined(dim_spec,keyString=KEY_UNGRIDDED_DIM_COORDINATES)
+            _ASSERT(.not.(has_units .and. has_coordinates), "Both extent and coordinates specified")
+            if (has_name) then
+               dim_name = ESMF_HConfigAsString(dim_spec, keyString=KEY_UNGRIDDED_DIM_NAME, _RC)
+            end if
```
Comment (tclune) [inline generic3g/ComponentSpecParser/parse_var_specs.F90:234]: Screams for a helper function ...

File: generic3g/OutputInfo.F90 [hunk 2]
```diff
@@ -223,7 +223,7 @@ function make_ungridded_dim(info, n, rc)
       type(UngriddedDim) :: make_ungridded_dim
       integer, intent(in) :: n
       type(ESMF_Info), intent(in) :: info
-      integer, optional, intent(out) :: rc 
+      integer, optional, intent(out) :: rc
       integer :: status
       character(len=:), allocatable :: key
```
Comment (tclune) [inline generic3g/OutputInfo.F90:228]: Looks unused.

File: generic3g/OutputInfo.F90 [hunk 3]
```diff
@@ -223,7 +223,7 @@ function make_ungridded_dim(info, n, rc)
       type(UngriddedDim) :: make_ungridded_dim
```
Comment (tclune) [inline generic3g/OutputInfo.F90:223]: This appears to be a bug.   The `result` is `dims`, but you put the value into the local `make_ungridded_dims` which is the name of the function ...

Comment (tclune) [reply generic3g/OutputInfo.F90:223]: Never mind.   Bad split in GitHub view ...

File: generic3g/specs/FieldSpec.F90 [hunk 4]
```diff
@@ -245,36 +246,36 @@ subroutine set_field_default(rc)
...
-               x_r8_2d = this%default_value   
+               x_r8_2d = this%default_value
             else if (rank == 3) then
                call ESMF_FieldGet(this%payload,farrayptr=x_r8_3d,_RC)
-               x_r8_3d = this%default_value   
+               x_r8_3d = this%default_value
             else if (rank == 4) then
                call ESMF_FieldGet(this%payload,farrayptr=x_r8_4d,_RC)
-               x_r8_4d = this%default_value   
+               x_r8_4d = this%default_value
```
Comment (tclune) [inline generic3g/specs/FieldSpec.F90:278]: Please open a ticket to have this somehow moved into "Darian's" rank/kind agnostic layer.  (If possible.)

**Review by tclune — APPROVED** [review]

File: generic3g/OutputInfo.F90 (hunk 2 above)
Comment (bena-nasa) [reply generic3g/OutputInfo.F90:228]: I didn't make this particular file, from previous PR. But that rc is used, there's lots of `_RC` in the procedure.

Comment (bena-nasa) [reply generic3g/OutputInfo.F90:228]: Used

Comment (bena-nasa) [reply generic3g/OutputInfo.F90:228]: I didn't make this particular file, from previous PR. But that rc is used, there's lots of `_RC` in the procedure.

File: generic3g/OutputInfo.F90 (hunk 3 above)
Comment (bena-nasa) [reply generic3g/OutputInfo.F90:223]: This is from previous PR, but yeah, looks like bug will fix

**Comment by tclune** [comment]
It was a different variable.   Since it was not part of your change – it did not properly identify.   But … as it turns out it was because GitHub was showing the 1st half of the procedure above and the latter half of the procedure below and the 2 had nearly identical names …


From: Ben Auer ***@***.***>
Date: Tuesday, July 16, 2024 at 9:27 AM
To: GEOS-ESM/MAPL ***@***.***>
Cc: Clune, Thomas L. (GSFC-6101) ***@***.***>, Review requested ***@***.***>
Subject: [EXTERNAL] [BULK] Re: [GEOS-ESM/MAPL] Implement ungridded dim mirroring and complete ungridded dim parsing (PR #2907)
CAUTION: This email originated from outside of NASA.  Please take care when clicking links or opening attachments.  Use the "Report Message" button to report suspicious messages to the NASA SOC.



@bena-nasa commented on this pull request.

________________________________

In generic3g/OutputInfo.F90<https://github.com/GEOS-ESM/MAPL/pull/2907#discussion_r1679397601>:

> @@ -223,7 +223,7 @@ function make_ungridded_dim(info, n, rc)

       type(UngriddedDim) :: make_ungridded_dim

       integer, intent(in) :: n

       type(ESMF_Info), intent(in) :: info

-      integer, optional, intent(out) :: rc

+      integer, optional, intent(out) :: rc

       integer :: status

       character(len=:), allocatable :: key

I didn't make this particular file, from previous PR. But that rc is used, there's lots of _RC in the procedure.

File: generic3g/specs/FieldSpec.F90 (hunk 4 above)
Comment (bena-nasa) [reply generic3g/specs/FieldSpec.F90:278]: I'll change this in this PR just so trivial to change.

Comment (bena-nasa) [reply generic3g/specs/FieldSpec.F90:278]: Well tried changing to use the function that should do this, but made test fail so I'll futz with this later.

**Review by tclune — APPROVED** [review]

**Pull request merged**